### PR TITLE
Support for BioSamples Submissions

### PR DESCRIPTION
**Pull request by MightyAx**
- [x] Accept CLI Arguments
- [x] Submit new Sample to BioSamples
- [x] Update Sample when `accession` provided
- [x] Provide `domain` CLI parameter
- [x] Ask for user input before submitting BioSamples that contain issues.
- [x] Bring units from spreadsheet to Sample Attribute (for geographic data)
- [x] ~~Populate 9606 ontology for `host_common_name`, `host_scientific_name`~~
    - Removed as BioSamples will auto curate these overnight.
- [x] Return friendlier error that `traverson` failing when accession is not found.
- [x] Report submission errors into the same objects as normal errors.

Out of Scope:
- Validate against BioSamples (without submitting)
    - Add feature to: [ebi-ait/python_biosamples-v4_lib](https://github.com/ebi-ait/python_biosamples-v4_lib)
- Return Accessions back to Excel (Better)

Dependant upon: [biosamples-v4#6](https://github.com/ebi-ait/python_biosamples-v4_lib/pull/6)
[Populated Test Sample](https://wwwdev.ebi.ac.uk/biosamples/samples/SAMEA5574070)
[Manually Populated Production Sample](https://www.ebi.ac.uk/biosamples/samples/SAMEA7098099), for comparison.

### CODE REVIEW

File: cli.py [hunk 1]
```diff
@@ -15,18 +18,95 @@ def write_dict(file_path, data_dict):
...
+                    # ToDo: Maybe append this to row['sample']['errors']
+                    # row['sample'].setdefault('errors', []).append()
+                    row.setdefault('sample_request', {})['error'] = f'Encoding Error: {error}'
+                    error_count = error_count + 1
         write_dict(json_file_path, data)
+        print(f'Data from {sample_count} BioSamples conversions and {error_count} errors written to: {json_file_path}')
+
+        biosample_count = 0
+        error_count = 0
+        for row in data:
```
Comment (ke4) [inline cli.py:96]: Do we really need this 2nd loop?
We could do all of this logic in the 1st I believe.

File: cli.py [hunk 2]
```diff
@@ -15,18 +18,95 @@ def write_dict(file_path, data_dict):
...
+                    # ToDo: Maybe store this in row['sample']['response']
+                    row['sample_response'] = biosamples.send_sample(row['sample_request'])
+                    biosample_count = biosample_count + 1
+                except Exception as error:
+                    # ToDo: Also write to issues
+                    # ToDo: Maybe append this to row['sample']['errors']
+                    # row['sample'].setdefault('errors', []).append()
+                    row.setdefault('sample_response', {})['error'] = f'BioSamples Error: {error}'
+                    error_count = error_count + 1
+        write_dict(json_file_path, data)
```
Comment (ke4) [inline cli.py:108]: This would overwrite the file that has been created in line 91

File: excel/validate.py [hunk 3]
```diff
@@ -28,17 +16,20 @@ def validate_object(object_validation: dict, object_data: dict):
...
                 mandatory = attribute_info['mandatory'].strip()
                 if mandatory == 'M':
-                    errors.append('Error: Mandatory Atrribute {} is missing.'.format(attribute_name))
+                    errors.append(f'Error: Mandatory Attribute {attribute_name} is missing.')
                 elif mandatory == 'R':
-                    errors.append('Warning: Reccomended Atrribute {} is missing.'.format(attribute_name))
+                    errors.append(f'Warning: Recommended Attribute {attribute_name} is missing.')
                 elif mandatory != 'O':
-                    errors.append('Info: Atrribute {} may be required. {}'.format(attribute_name, mandatory))
+                    errors.append(f'Info: Attribute {attribute_name} may be required. {mandatory}')
```
Comment (ke4) [inline excel/validate.py:23]: Do we need `{mandatory}` at the end of this message?

File: services/biosamples.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def optional_attribute(sample: dict, attribute: str):
+    return sample[attribute] if object_has_attribute(sample, attribute) else None
+
+
+def make_attribute(name, value) -> Attribute:
+    units = None
+    # ToDo: Refactor to create the Sample at validation-time so the units are accessible from validation_map
+    if name in ['geographic_location_(latitude)', 'geographic_location_(longitude)']:
```
Comment (ke4) [inline services/biosamples.py:20]: These are `magic strings`. Is it not better to put them into a config file and get them from there?

File: services/biosamples.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    attributes = []
+    for name, value in sample.items():
+        if value_populated(value):
+            attributes.append(make_attribute(name, value))
+    return attributes
+
+
+def fixup_sample(sample: dict) -> dict:
+    fixed_sample = sample.copy()
+    remove_keys = ['errors', 'schema_erros', 'sample_accession', 'sample_alias', 'sample_title', 'sample_description', 'tax_id',
```
Comment (ke4) [inline services/biosamples.py:35]: These values could go to a config file, too

**Review by ke4 — CHANGES_REQUESTED** [review]
I added some minor change requests, otherwise it is looking good.

File: cli.py (hunk 1 above)
Comment (MightyAx) [reply cli.py:96]: yep, we can probably refactor away some of the current oddities. It was coded very defensively.

File: cli.py (hunk 2 above)
Comment (MightyAx) [reply cli.py:108]: Yep, that is intended, to update the file with the latest information.

File: excel/validate.py (hunk 3 above)
Comment (MightyAx) [reply excel/validate.py:23]: yep: for example: 'M for paired fastq; NA for single fastq'

File: services/biosamples.py (hunk 4 above)
Comment (MightyAx) [reply services/biosamples.py:20]: Once the ToDo is implemented this won't be an issue.

File: services/biosamples.py (hunk 5 above)
Comment (MightyAx) [reply services/biosamples.py:35]: You could but that would make the code much harder to follow.
Unless we implemented some mapping of excel column names to names for BioSamples (or other services).

**Pull request merged**